[FIX] core: extract normalize from HTML sanitizer

rdeodoo opened the pull request:
Since [1], it's possible to conditionnally bypass the HTML sanitizer in field definition with the `sanitize_overridable` attribute. In a nutshell, when someone is part of the required group(s), it won't go through the sanitizer, while the people not part of the group(s) will.

A behavior was thus introcuded to prevent a "restricted" user to wipe the changes done previously by an "elevated" user (which bypassed the sanitizer). But that behavior was not correct as there was unforeseen cases which led to raise this error which are not due to the sanitizer but to normalization.

Indeed, while named `html_sanitize()`, it also does some normalize stuff on top of the real sanitize part.
For instance, there is also (not exhaustive):
- some MAKO compatibility, replacing some chars
- special case for quotes, related to mail clients, which will add data attributes, add nodes in dom etc. This happen when the following are found:
  - `<blockquote/>` tag
  - text-based quotes (>, >>) and signatures (-- Signature)
  - html signature (-- <br />blah)
- some editor compatibility which removed the wrapping `<div/>` element
- `nbsp` handling..

See commit list below for detail about how/when/why those normalize cases where introduced.

At the end, the issue was that the normalize part should not prevent a "restricted" user to modify the content of an "elevated" user. Only the sanitize part should.

For instance, the `Quotes` snippet dropped by an "elevated" user was preventing further edition by a "restricted" user because there was a "false positive" raised when checking if the save would wipe the existing changes. Indeed, when the "elevated" user droped the snippet, it was saved as:
```html
<blockquote class=".." data-name="Blockquote">
```
But when the "restricted" user then wanted to do some changes, it would become:
```html
<blockquote class=".." data-name="Blockquote" data-o-mail-quote-node="1" data-o-mail-quote="1">
```

Same for `Share` snippet:
```html
<a href="https://www.facebook.com/sharer/sharer.php?u={url}">
<a href="https://www.facebook.com/sharer/sharer.php?u=%7Burl%7D">
```

[1]: https://github.com/odoo/odoo/commit/cf844e34dd0ce4830eb99fd0fa5b6b9cb58c867c

Normalize commit list:
https://github.com/odoo/odoo/commit/5f1ec49ecdac6d72cd42755c41fbe75d6a1f3587 https://github.com/odoo/odoo/commit/69af79ff3d705d19a71ba3ba7851b981cb301077 https://github.com/odoo/odoo/commit/2bcf4cca79a57dfba84d1f3e3fa7b8908bfe66e8 https://github.com/odoo/odoo/commit/f5688cd8fd515d1b668e8eb1d74de68faa681a01 https://github.com/odoo/odoo/commit/cb8c2d2b7e15c7c16e02d078767e27a07e5012c6 https://github.com/odoo/odoo/commit/275ee5825d38841a3eb21bb195722f3ceed09005 https://github.com/odoo/odoo/commit/b51d21c5b83b88e8d56dbbbb7600bcbe554d1b07

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/109939).

rdeodoo commented on odoo/fields.py:
```diff
@@ -1979,27 +1980,30 @@ def _convert(self, value, record, validate):
             'strip_classes': self.strip_classes
         }
 
+        # TODO: Wouldn't it be better to move all the normalize stuff here
```
I meant before the `return value` a few lines above, so all call to `_convert()` does it.
And so even when saved by the admin, the normalize part will be done.
Isn't it strange to have this conditioned? Having that based on right or on field attribute (sanitize) doesn't seem right?
Eg, why would we remove the wrapping `<div/>` element only on sanitized fields?

rdeodoo commented on odoo/tools/mail.py:
```diff
@@ -177,20 +105,140 @@ def parse_style(self, el):
+        if doc is not None:
+            for el in doc.iter(tag=etree.Element):
+                tag_quote(el)
+    # TODO: Those `except` cases are copy/paster from `html_sanitize`, fix that
```
Waiting for first review before tackling this one.
I am not sure the "empty" case is actually still needed in the other place.

rdeodoo reviewed (commented):
Here is a first version of it. The runbot is green which is already something given the strong test coverage about those mail/sanitize flows.

I am not sure I could have made the diff easier to read, but note that the `tag_quote()` diff is just a 100% move (so are most of the rest of the lines on their own).

Any review welcome here.

I'll write some tests to cover the cases that are fixed here (as described in commit message) in the meantime.

cc @JKE-be

JKE-be commented on odoo/fields.py:
```diff
@@ -1979,27 +1980,30 @@ def _convert(self, value, record, validate):
 
-            original_value = record[self.name]
+            original_value = html_normalize_post(html_normalize_pre(record[self.name]))
             if original_value:
-                initial_value_sanitized = html_sanitize(original_value, **sanitize_vals)
+                original_value_sanitized = html_sanitize(original_value, **sanitize_vals)
```
so in this case, you will have a second call to _pre and _post normalize via the call to html_sanitize :thinking: and still risk to compare apple and orange. no ?

line 1991/2000 
```py
original_value = record[self.name])
get_parsed(html_normalize_post(html_normalize_pre(original_value)) != ...
```

rdeodoo commented on odoo/tools/mail.py:
```diff
@@ -177,19 +105,125 @@ def parse_style(self, el):
+            if doc is not None:
+                for el in doc.iter(tag=etree.Element):
+                    tag_quote(el)
+        # TODO: Those `except` cases are copy/paster from `html_sanitize`, fix that
```
Waiting for first review before tackling this one (probably going to use `@wraps` here?).
(I am not sure the "empty" case is actually still needed in the other place.)
Probably someone will have something to say about `def html_normalize(src, pre_sanitize=True, post_sanitize=True, silent=True):` implementation, following that I'll adapt.

rdeodoo replied:
Ok, I misunderstood what you meant by 

> But fundamentally the entire silent feature seems unnecessary to maintain?

Adapted the way you meant it.

rdeodoo commented on odoo/tools/mail.py:
```diff
@@ -177,19 +105,125 @@ def parse_style(self, el):
+        el.set('data-o-mail-quote', '1')
+
+
+def html_normalize(src, pre_sanitize=True, post_sanitize=True, silent=True):
```
Is [this](https://github.com/odoo/odoo/pull/109939/commits/7df78c1907df595d850714316485526100b49055?w=1) (mind the `?w=1` in URL) what you had in mind?

rdeodoo replied:
Seems about right, the only exception that will be handled ( and in a "silently" way) is the empty document parser error, the rest will be up to the caller to handle :+1:

rdeodoo commented on odoo/tools/mail.py:
```diff
@@ -177,19 +105,125 @@ def parse_style(self, el):
+            doc = html.fromstring(src)
+            if doc is not None:
+                for el in doc.iter(tag=etree.Element):
+                    tag_quote(el)
```
Seems about right, the only exception that will be handled ( and in a "silently" way) is the empty document parser error, the rest will be up to the caller to handle :+1:

rdeodoo commented on odoo/fields.py:
```diff
@@ -1985,21 +1986,22 @@ def _convert(self, value, record, validate):
+                original_value_normalized = html_normalize(original_value)
 
                 def get_parsed(val):
                     return etree.tostring(html.fromstring(val))
```
Indeed, nice catch..
Note that I made a test for this use case back then: https://github.com/odoo/odoo/blame/1a01c8470b1971dfc2b06dcce0e06fe969ac8040/odoo/addons/test_new_api/tests/test_new_fields.py#L3064-L3075
So it's even tested and indeed not needed anymore as both checked values are going through the same code (fromstring & tostring) already in `html_normalize()` now.

rdeodoo commented on odoo/tools/mail.py:
```diff
@@ -177,82 +105,168 @@ def parse_style(self, el):
+            cleaned = cleaner.clean_html(doc)
+        except etree.ParserError as e:
+            if 'empty' in str(e):
+                return u""
```
:+1:

I also wonder if we should really keep the bare `Exception` case, it seems to have been introduced alongside those commits:
- [d7185be](https://github.com/odoo/odoo/commit/d7185be38c2d386c1826d7e0aefacf15828d7050#diff-b583c3495a9c5bf0e224788074ddaabf14647a6e2b8cca1fe273492986b187acR57) (not yet)
- [a74d50d](https://github.com/odoo/odoo/commit/a74d50d220b26f26a37faeb59e8923f1e635c2ea#diff-b583c3495a9c5bf0e224788074ddaabf14647a6e2b8cca1fe273492986b187acR55) (not yet, double exception added because of retrying)
- [f97bd8b](https://github.com/odoo/odoo/commit/f97bd8bd6313db276452df1e4be176fa0d2ce292#diff-b583c3495a9c5bf0e224788074ddaabf14647a6e2b8cca1fe273492986b187acR65) (here)
- [6fd552c](https://github.com/odoo/odoo/commit/6fd552c3ceae47384aca1ac960cb3fd5f8bbb832#diff-b583c3495a9c5bf0e224788074ddaabf14647a6e2b8cca1fe273492986b187acR91) (fine tuned)

I guess it's anyway not worth the risk in 16.0

rdeodoo commented on odoo/tools/mail.py:
```diff
@@ -177,82 +105,168 @@ def parse_style(self, el):
+            # - HTML comment only string like `<!-- abc -->`
+            # - whitespace only string
+            cleaner = _Cleaner(**kwargs)
+            cleaned = cleaner.clean_html(doc)
```
I am not sure to get this one. Could you explain? Without the `clean_html()` part it's returning nothing

rdeodoo reviewed (commented):
@xmo-odoo Changes applied [here](https://github.com/odoo/odoo/compare/706b4829eda5bc159aecd7e0437ca219f9c35e5f..9fd9c84b933d3b94435faba592d91c57b3569239)

rdeodoo commented:
@robodoo rebase-ff

Changes applied:
- removed the `clean_html()` call, useful links:
  - https://lxml.de/api/lxml.html.clean-pysrc.html#Cleaner.__call__
  - https://lxml.de/api/lxml.html-pysrc.html#_transform_result
  - https://lxml.de/api/lxml.html.clean-pysrc.html#Cleaner.clean_html
- removed the first REV commit and adapted the line with this refactoring as it was actually fix with https://github.com/odoo/odoo/pull/110148 a day after

Thank you for the review(s) @xmo-odoo 

@qsm-odoo In case you want to have a look, otherwise I'll self r+

robodoo commented:
Merge method set to rebase and fast-forward.

rdeodoo commented:
Thanks for the review(s).
Last [changes](https://github.com/odoo/odoo/compare/e67e3b2a159a11279099129d39b3d49a4f4447a1..23318d585504572421100d606ae207c74b3463b6): applied last ODO suggestions & adapted the .pot accordingly

@robodoo r+

rdeodoo commented:
@robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.

rdeodoo commented on odoo/tools/mail.py:
```diff
@@ -177,82 +105,165 @@ def parse_style(self, el):
+        el.set('data-o-mail-quote', '1')
+
+
+def html_normalize(src, callback=None):
```
Thanks for writing that, went with `filter_callback`
